include/oneapi/dpl/internal/histogram_impl.h

```diff
@@ -0,0 +1,116 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_HISTOGRAM_IMPL_H
+#define _ONEDPL_HISTOGRAM_IMPL_H
+
+#include "function.h"
+#include "histogram_extension_defs.h"
+#include "histogram_binhash_utils.h"
+#include "../pstl/iterator_impl.h"
+
+#if _ONEDPL_HETERO_BACKEND
+#    include "oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h"
+#endif
+
+namespace oneapi
+{
+namespace dpl
+{
+
+namespace __internal
+{
+
+//Only supported for dpcpp backend
+#if _ONEDPL_BACKEND_SYCL
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _IdxHashFunc,
+          typename _RandomAccessIterator2>
+inline oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
+    {
+        oneapi::dpl::__par_backend_hetero::__parallel_histogram(::std::forward<_ExecutionPolicy>(exec), __first, __last,
+                                                                __num_bins, __func, __histogram_first);
+    }
+}
+#endif // _ONEDPL_BACKEND_SYCL
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _IdxHashFunc,
+          typename _RandomAccessIterator2>
+inline oneapi::dpl::__internal::__enable_if_host_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    static_assert(sizeof(_Size) == 0 /*false*/, "Histogram API is not currently supported for host parallel policies");
```

Comment:
The intention is to add this implementation at a later time.
For now, I think that this makes things most clear that we are intentionally leaving out the CPU implementation of this API, rather than just a compilation error about a missing function definition for `__pattern_histogram` if a user tries to call it with with a host parallel policy.
  
I don't feel very strongly on this topic though, so I'm open to other options. Do you have a preference to just remove this definition or explicitly delete it?